requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c

```diff
@@ -0,0 +1,98 @@
+/* Body file for function EventBus
+ * Generated by TASTE (kazoo/templates/skeletons/c-body/function.tmplt)
+ * You can edit this file, it will not be overwritten
+
+    !! IMPORTANT                                                        !!
+    !! When you modify your design (interface view), you must update    !!
+    !! the procedures corresponding to the provided interfaces in this  !!
+    !! file. The up-to-date signatures can be found in the .ads file.   !!
+*/
+#include "eventbus.h"
+#include "routing.h"
+
+#define SUBSCRIPTION_TABLE_BUFFER_SIZE 256
+
+typedef struct {
+    asn1SccPID pid;
+    asn1SccT_UInt32 eventId;
+} Subscription;
+
+static int subscriptionCounter;
+static Subscription subscriptionTable[SUBSCRIPTION_TABLE_BUFFER_SIZE] = {{0}};
+
+
+int eventbus_getSubscriptionIndex(const asn1SccPID pid, const asn1SccT_UInt32 eventId)
+{
+    for(int i = 0; i < subscriptionCounter || i < SUBSCRIPTION_TABLE_BUFFER_SIZE; i++)
+    {
+        if(subscriptionTable[i].pid == pid && subscriptionTable[i].eventId == eventId)
+        {
+            return i;
+        }
+    }
+    return -1; // Subscription is not present
+}
+
+void eventbus_removeEmptySubscription(int emptySubscriptionIndex)
+{
+    for (int j = emptySubscriptionIndex; j < subscriptionCounter || j < SUBSCRIPTION_TABLE_BUFFER_SIZE - 1; j++)
+    {
+         subscriptionTable[j] = subscriptionTable[j + 1];
+    }
+}
+
+void eventbus_updateRoutingTable(asn1SccT_UInt32 eventId)
+{
+    for(int i = 0; i<subscriptionCounter; i++)
+    {
+        if(subscriptionTable[i].eventId == eventId)
+        {
+            eventbus_dispatch_event_routing_table[subscriptionTable[i].pid] = true;
+        }
+    }
+}
+
+void eventbus_startup(void)
+{
+    subscriptionCounter = 0;
+    eventbus_dispatch_event_set_all_routes_enabled(false);
+}
+
+void eventbus_PI_receive_event
+      (const asn1SccT_EventMessage *IN_eventmessage)
+
+{
+    eventbus_updateRoutingTable(IN_eventmessage->id);
+
+    eventbus_RI_dispatch_event(IN_eventmessage);
+
+    eventbus_dispatch_event_set_all_routes_enabled(false);
```

Comment:
Not exactly the fastest solution. In the future please consider using e.g., 2D array. While the current solution is more portable, use of C++ could be considered (hashmaps for better memory utilization). updateRoutingTable could be modified to turn the routes both on and off, depending on the event. Resetting all to enabled wouldn't be necessary then.

Reply:
Done